diff -ruN sqsh-2.1.5.orig/configure.in sqsh-2.1.5/configure.in
--- sqsh-2.1.5.orig/configure.in	2008-12-18 11:42:02.000000000 -0600
+++ sqsh-2.1.5/configure.in	2008-12-18 11:42:41.000000000 -0600
@@ -51,6 +51,7 @@
 		then
 			MOTIF_PATH=$with_motif
 			with_motif="yes"
+			with_x="yes"
 			MOTIF_INCDIR="-I$MOTIF_PATH/include/X11"
 			MOTIF_LIBDIR="-L$MOTIF_PATH/lib"
 		else
@@ -58,7 +59,6 @@
 			MOTIF_LIBDIR="# -L/usr/X11/lib"
 		fi
 
-		with_x="yes"
 		DEF_MOTIF="-DUSE_MOTIF"
 		AC_SUBST(MOTIF_PATH)
 		AC_SUBST(MOTIF_LIBS)
