diff -ruN sqsh-2.1.5.orig/configure.in sqsh-2.1.5/configure.in
--- sqsh-2.1.5.orig/configure.in	2008-12-18 12:08:48.000000000 -0600
+++ sqsh-2.1.5/configure.in	2008-12-18 12:10:00.000000000 -0600
@@ -55,6 +55,7 @@
 			DEF_MOTIF="-DUSE_MOTIF"
 			MOTIF_PATH=$with_motif
 			with_motif="yes"
+			with_x="yes"
 			MOTIF_INCDIR="-I$MOTIF_PATH/include/X11"
 			MOTIF_LIBDIR="-L$MOTIF_PATH/lib"
 		else
@@ -63,7 +64,6 @@
 			MOTIF_LIBDIR="# -L/usr/X11/lib"
 		fi
 
-		with_x="yes"
 	],[
 		DEF_MOTIF="# -DUSE_MOTIF"
 		with_motif="no"
