diff -Naur gimp-2.6.0.orig/app/config/gimpguiconfig.c gimp-2.6.0/app/config/gimpguiconfig.c
--- gimp-2.6.0.orig/app/config/gimpguiconfig.c	2008-09-30 23:21:55.000000000 +0200
+++ gimp-2.6.0/app/config/gimpguiconfig.c	2008-10-01 19:13:30.574438763 +0200
@@ -45,7 +45,7 @@
 #elif HAVE_CARBON
 #  define DEFAULT_WEB_BROWSER  "open %s"
 #else
-#  define DEFAULT_WEB_BROWSER  "firefox %s"
+#  define DEFAULT_WEB_BROWSER  "xdg-open %s"
 #endif
 
 
diff -Naur gimp-2.6.0.orig/docs/gimprc.5.in gimp-2.6.0/docs/gimprc.5.in
--- gimp-2.6.0.orig/docs/gimprc.5.in	2008-09-29 20:46:42.000000000 +0200
+++ gimp-2.6.0/docs/gimprc.5.in	2008-10-01 19:13:59.214001847 +0200
@@ -788,7 +788,7 @@
 web-browser.
 
 .TP
-(web-browser "firefox %s")
+(web-browser "xdg-open %s")
 
 Sets the external web browser to be used.  This can be an absolute path or the
 name of an executable to search for in the user's PATH. If the command
diff -Naur gimp-2.6.0.orig/etc/gimprc gimp-2.6.0/etc/gimprc
--- gimp-2.6.0.orig/etc/gimprc	2008-09-29 20:46:42.000000000 +0200
+++ gimp-2.6.0/etc/gimprc	2008-10-01 19:14:26.252070783 +0200
@@ -625,7 +625,7 @@
 # appended to the command with a space separating the two.  This is a single
 # filename.
 # 
-# (web-browser "firefox %s")
+# (web-browser "xdg-open %s")
 
 # When enabled, the online user manual will be used by the help system.
 # Otherwise the locally installed copy is used.  Possible values are yes and
